src/centered-on-asteroid/utils.py

```diff
@@ -5,15 +5,260 @@
+
+    # Calculate the maximum difference in gradients
+    max_grad_diff = max(grad_diffs)
+    return torch.Tensor([max_grad_diff])
+
+
+def get_angles(positions: torch.Tensor, epsilon: float = 1e-6) -> torch.Tensor:
+    """
+    Returns the angles between the 4 images
+
+    Args:
+        positions (torch.Tensor): (x, y) position for the 4 images. Shape: (8)
+
+    Returns: Angles (3,)
+    """
+    deltas = get_movement_vectors(positions)
+
+    angles = []
+    for i in range(0, len(deltas)):
+        if deltas[i][0] < 0:
+            angle = torch.atan(deltas[i][1] / (deltas[i][0] - epsilon))
+        else:
+            angle = torch.atan(deltas[i][1] / (deltas[i][0] + epsilon))
+        angles.append(angle * 180 / math.pi)
+
+        if angles[-1].isnan():
+            print(f"{deltas[i][0]}, {deltas[i][1]}")
+            print(f"NaN angle detected: {angles[-1]}")
+            print(positions)
+        if angles[-1].isinf():
+            print(f"Infinite angle detected: {angles[-1]}")
+    return torch.stack(angles)
+
+
+def get_max_ang_diff(positions: torch.Tensor) -> torch.Tensor:
+    """
+    Returns the maximum difference in angle between the 4 images
+
+    Args:
+        positions (torch.Tensor): (x, y) position for the 4 images. Shape: (8)
+
+    Returns: Maximum difference in angle (1,)
+    """
+    angles = get_angles(positions)
+
+    ang_diffs = []
+    for ang1, ang2 in torch.combinations(angles, r=2, with_replacement=False):
+        diff = abs(ang1 - ang2)
+        if diff > 180:
+            diff = 360 - diff
+        ang_diffs.append(diff)
+
+    return torch.Tensor([max(ang_diffs)])
+
+
+def get_movement_vectors(positions: torch.Tensor) -> torch.Tensor:
+    """
+    Returns dx, dy movement vectors between the 4 images
+
+    Args:
+        positions (torch.Tensor): (x, y) position for the 4 images. Shape: (8,)
+
+    Returns: Movement vectors (3, 2)
+    """
+    # Reshape to (4, (x, y))
+    positions = positions.view(4, 2)
+
+    deltas = []
+    for i in range(1, len(positions)):
+        deltas.append(positions[i] - positions[i - 1])
+
+    return torch.stack(deltas)
+
+
+def get_dataframe(real_movers_csv: str, bogus_movers_csv: str) -> pd.DataFrame:
     """
     Returns a DataFrame object grouped by the mover_id with columns file_name and label
+
+    Args:
+        real_movers_csv (str): Path to the csv file containing the real movers
+        bogus_movers_csv (str): Path to the csv file containing the bogus movers
     """
     # Read csv
-    real_movers = pd.read_csv(path_to_csvs + "movers_images_lookup.csv")
-    bogus_movers = pd.read_csv(path_to_csvs + "rejected_movers_images_lookup.csv")
+    real_movers = pd.read_csv(real_movers_csv)
+    bogus_movers = pd.read_csv(bogus_movers_csv)
+
+    # Ignore rows with missing data
+    real_movers = real_movers.dropna(subset=[x_cord, y_cord, "file_name"])
+    bogus_movers = bogus_movers.dropna(subset=[x_cord, y_cord, "file_name"])
```

Comment:
This should be optional because we only want to filter if we need them for the feature vectors